src/porepy/examples/run_profiling.py

```diff
@@ -0,0 +1,286 @@
+        model_params["refinement_level"] = args.grid_refinement
+    else:
+        raise ValueError(f"{args.grid_refinement=}")
+
+    # Select a model based on choice of physics and geometry.
+    model: Optional[Type] = None
+    if args.geometry == 0:
+        if args.physics == "flow":
+            model = FlowBenchmark2dCase1Model
+        elif args.physics == "poromechanics":
+            model = Case1Poromech2D
+    elif args.geometry == 1:
+        if args.physics == "flow":
+            model = FlowBenchmark2dCase3aModel
+        elif args.physics == "poromechanics":
+            model = Case3aPoromech2D
+    elif args.geometry == 2:
+        if args.physics == "flow":
+            model = FlowBenchmark2dCase4Model
+        elif args.physics == "poromechanics":
+            model = Case4Poromech2D
+    elif args.geometry == 3:
+        if args.physics == "flow":
+            model = FlowBenchmark3dCase3Model
+        elif args.physics == "poromechanics":
+            model = Case3Poromech3D
+
+    if model is None:
+        raise ValueError(f"{args.geometry=}, {args.physics=}")
+
+    return model(model_params)
+
+
+def run_model_with_tracer(args, model) -> None:
+    """Run a model with VizTracer enabled for performance profiling.
+
+    Parameters:
+        args: Command-line arguments containing the following attributes:
+            - physics (str): The physics of the model.
+            - geometry (str): The geometry of the model.
+            - grid_refinement (int): The grid refinement level for the model.
+            - save_file (str): The file path to save the profiling results. If empty, a
+            default name is generated based on chosen physics, geometry, and grid
+            refinement.
+            - min_duration (int): Minimum duration in microseconds for a function to be
+            recorded by VizTracer.
+            - keep_output (bool): Whether to keep the output file after viewing it.
+        model: The model to be run and profiled.
+
+    Raises:
+        ValueError: If ``args.save_file`` does not end in .json.
+
+    Returns:
+        None
+
+    """
+    if args.save_file == "":
+        save_file: str = (
+            f"profiling_{args.physics}_{args.geometry}_{args.grid_refinement}.json"
+        )
+    else:
+        if not args.save_file.endswith(".json"):
+            raise ValueError(f"{args.save_file=}")
+        save_file = args.save_file
+
+    # Run model with viztracer enabled.
+    tracer = VizTracer(
+        min_duration=args.min_duration,  # μs
+        ignore_c_function=True,
+        ignore_frozen=True,
+    )
+    tracer.start()
+    model.prepare_simulation()
+    print("Num dofs:", model.equation_system.num_dofs())
+    pp.run_time_dependent_model(
+        model,
+        {
+            "prepare_simulation": False,
+            "nl_divergence_tol": 1e8,
+            "max_iterations": 25,
+            "nl_convergence_tol": 1e-2,
+            "nl_convergence_tol_res": 1e-2,
```

Comment:
These are very loose tolerances. Any reason to do so?

Reply:
I see two dangers:
1. This will underestimate the time spent on solvers compared to everything else.
2. There is a risk that a loose tolerance may lead the solution to a weird part of the solution space, thus to profiling results that are biased in some sense.

Reply:
I decided to make it like this because otherwise I could not make Newton converge for some of the models. With this, it takes 1-2 Newton iterations to accept the solution. The whole simulation consists of only one time step, so we don't use these unphysical solutions for the next time step. I'm also afraid of making the benchmark too time consuming if forcing it to make 10 or more Newton iterations.

I see your point about biased results, but I can't fully agree with it. As I understand, we are now interested in measuring the times of: 1) initialization 2) assembly 3) linear and nonlinear solver. 1 and 2 should be identical for any physical state and material constants. For a linear solver, it should be more or less the same for a direct solver. What matters for iterative linear solvers and Newton is the number of iterations, while each iteration should take the same time and also not depend on physical state.

Anyway, to make this work properly with the correct convergence tolerances, one should find the proper time step and the physical material constants and initial and boundary conditions, which I've not done, but I can do it if needed.

Reply:
I see your point in differing between simulations for benchmarking and actual example simplifications. Let's keep this for now, but please explain in comments what is being done and the rationale behind it.

Reply:
Added a comment